File: app/Support/StudySubjectCards.php
<?php

namespace App\Support;

use Carbon\Carbon;
use Illuminate\Support\Collection;

class StudySubjectCards
{
    public function build(Collection $studySubjects): array
    {
        $maxSeconds = max(1, (int) $studySubjects->max('duration_seconds_total'));
        $tones = ['violet', 'cyan', 'amber', 'emerald', 'indigo'];
        $icons = ['book', 'code', 'database', 'language', 'target'];

        return $studySubjects
            ->values()
            ->map(function ($subject, int $index) use ($maxSeconds, $tones, $icons) {
                $seconds = (int) ($subject->duration_seconds_total ?? 0);

                return [
                    'id' => $subject->id,
                    'name' => $subject->name,
                    'description' => $subject->description,
                    'photo_url' => $subject->photo_path ? asset('storage/'.$subject->photo_path) : null,
                    'seconds' => $seconds,
                    'hours_label' => $this->formatStudySecondsAsHours($seconds),
                    'goal_period' => $subject->goal_period,
                    'goal_minutes' => $subject->goal_minutes,
                    'goal_value' => $subject->goal_minutes ? $this->formatGoalValue((int) $subject->goal_minutes) : '',
                    'goal_unit' => $subject->goal_minutes && $subject->goal_minutes % 60 === 0 ? 'hours' : 'minutes',
                    'goal_progress' => $this->buildSubjectGoalProgress($subject, $seconds, $maxSeconds),
                    'goal_label' => $this->buildSubjectGoalLabel($subject),
                    'recent_at' => $this->recentActivityDate($subject)?->toDateTimeString(),
                    'tone' => $tones[$index % count($tones)],
                    'icon' => $icons[$index % count($icons)],
                ];
            })
            ->all();
    }

    public function sortByRecentActivity(Collection $studySubjects): Collection
    {
        return $studySubjects
            ->sortByDesc(fn ($subject) => $this->recentActivityDate($subject)?->timestamp ?? 0)
            ->values();
    }

    private function recentActivityDate($subject): ?Carbon
    {
        $dates = collect([
            $subject->updated_at,
            $subject->created_at,
            $subject->last_studied_at,
        ])->filter();

        if ($dates->isEmpty()) {
            return null;
        }

        return $dates
            ->map(fn ($date) => Carbon::parse($date))
            ->sortByDesc(fn (Carbon $date) => $date->timestamp)
            ->first();
    }

    private function buildSubjectGoalProgress($subject, int $totalSeconds, int $maxSeconds): int
    {
        if (! $subject->goal_minutes) {
            return $totalSeconds > 0 ? (int) round(($totalSeconds / $maxSeconds) * 100) : 0;
        }

        $periodSeconds = $subject->goal_period === 'weekly'
            ? (int) ($subject->duration_seconds_week ?? 0)
            : (int) ($subject->duration_seconds_today ?? 0);

        return min(100, (int) round((intdiv($periodSeconds, 60) / max(1, (int) $subject->goal_minutes)) * 100));
    }

    private function buildSubjectGoalLabel($subject): string
    {
        if (! $subject->goal_minutes) {
            return 'Sem meta definida';
        }

        $period = $subject->goal_period === 'weekly' ? 'semana' : 'dia';

        return 'Meta: '.$this->formatMinutesAsHours((int) $subject->goal_minutes).' por '.$period;
    }

    private function formatGoalValue(int $minutes): string
    {
        if ($minutes % 60 === 0) {
            return (string) intdiv($minutes, 60);
        }

        return (string) $minutes;
    }

    private function formatStudySecondsAsHours(int $seconds): string
    {
        $minutes = intdiv($seconds, 60);

        if ($seconds > 0 && $minutes === 0) {
            $minutes = 1;
        }

        return $this->formatMinutesAsHours($minutes).' estudadas';
    }

    private function formatMinutesAsHours(int $minutes): string
    {
        $hours = intdiv($minutes, 60);
        $remaining = $minutes % 60;

        if ($hours === 0) {
            return $remaining.'m';
        }

        if ($remaining === 0) {
            return $hours.'h';
        }

        return $hours.'h'.$remaining;
    }
}

File: app/services/study_subject_cards.rb
class StudySubjectCards
  def build(study_subjects)
    max_seconds = [study_subjects.map(&:duration_seconds_total).max.to_i, 1].max
    tones = %w[violet cyan amber emerald indigo]
    icons = %w[book book book book target]

    study_subjects.each_with_index.map do |subject, index|
      seconds = subject.duration_seconds_total.to_i

      {
        id: subject.id,
        name: subject.name,
        description: subject.description,
        photo_url: subject.photo.attached? ? Rails.application.routes.url_helpers.rails_blob_path(subject.photo, only_path: true) : subject.photo_path.present? ? "/storage/#{subject.photo_path}" : nil,
        seconds: seconds,
        hours_label: "#{format_study_seconds_as_hours(seconds)} estudadas",
        goal_minutes: subject.goal_minutes,
        goal_value: subject.goal_minutes.present? ? format_goal_value(subject.goal_minutes) : "",
        goal_unit: subject.goal_minutes.present? && (subject.goal_minutes % 60).zero? ? "hours" : "minutes",
        goal_progress: build_subject_goal_progress(subject, seconds, max_seconds),
        goal_label: build_subject_goal_label(subject),
        recent_at: recent_activity_date(subject)&.to_datetime&.to_s,
        tone: tones[index % tones.length],
        icon: icons[index % icons.length]
      }
    end
  end

  def sort_by_recent_activity(study_subjects)
    study_subjects.sort_by { |subject| -(recent_activity_date(subject)&.to_i || 0) }
  end

  private

  def recent_activity_date(subject)
    [subject.updated_at, subject.created_at, subject.last_studied_at].compact.max
  end

  def build_subject_goal_progress(subject, total_seconds, max_seconds)
    if subject.goal_minutes.blank?
      return total_seconds.positive? ? ((total_seconds.to_f / max_seconds) * 100).round : 0
    end

    period_seconds = subject.duration_seconds_week.to_i
    [((period_seconds / 60.0) / [subject.goal_minutes.to_i, 1].max * 100).round, 100].min
  end

  def build_subject_goal_label(subject)
    return "Sem meta definida" if subject.goal_minutes.blank?

    "Meta: #{format_minutes_as_hours(subject.goal_minutes)} por semana"
  end

  def format_goal_value(minutes)
    minutes % 60 == 0 ? (minutes / 60).to_s : minutes.to_s
  end

  def format_study_seconds_as_hours(seconds)
    minutes = seconds.to_i / 60
    minutes = 1 if seconds.positive? && minutes.zero?
    format_minutes_as_hours(minutes)
  end

  def format_minutes_as_hours(minutes)
    hours = minutes / 60
    remaining = minutes % 60
    return "#{remaining}m" if hours.zero?
    return "#{hours}h" if remaining.zero?

    "#{hours}h#{remaining}"
  end
end
